File: core/src/commonMain/kotlin/io/motionengine/core/Features.kt
package io.motionengine.core

import kotlin.math.*

data class PoseFeatures(val timestampMs: Long, val values: FloatArray) {
    val knee: Float get() = (values[0] + values[1]) * 90f
    val elbow: Float get() = (values[4] + values[5]) * 90f
    val torsoTilt: Float get() = values[6] * 180f
    val hipY: Float get() = values[10]
    val footY: Float get() = values[11]
    val imageScale: Float get() = values[12]
}

/** Feature v1: 16 base measurements followed by their 16 per-second derivatives. */
class FeatureExtractor {
    private var previous: PoseFeatures? = null

    fun reset() { previous = null }

    /** null means visible joints/geometry are insufficient. Invalid input throws before state changes. */
    fun extract(frame: PoseFrame): PoseFeatures? {
        frame.validate()
        val base = base(frame)
        if (base == null) { reset(); return null }
        val values = FloatArray(FEATURE_COUNT)
        base.copyInto(values)
        val prev = previous
        val dt = prev?.let { (frame.timestampMs - it.timestampMs) / 1000f } ?: 0f
        if (prev != null && dt > 0f && dt <= 0.5f) {
            for (i in 0 until BASE_COUNT) values[BASE_COUNT + i] = ((base[i] - prev.values[i]) / dt).coerceIn(-20f, 20f)
        }
        return PoseFeatures(frame.timestampMs, values).also { previous = it }
    }

    companion object {
        const val VERSION = 1
        const val BASE_COUNT = 16
        const val FEATURE_COUNT = 32
        val requiredJoints = listOf(11, 12, 13, 14, 15, 16, 23, 24, 25, 26, 27, 28)

        private fun midpoint(a: Landmark, b: Landmark) = Landmark((a.x + b.x) / 2, (a.y + b.y) / 2, (a.z + b.z) / 2)
        private fun distance(a: Landmark, b: Landmark): Float = sqrt((a.x-b.x).pow(2) + (a.y-b.y).pow(2) + (a.z-b.z).pow(2))
        private fun angle(a: Landmark, b: Landmark, c: Landmark): Float? {
            val ab = distance(a, b); val cb = distance(c, b)
            if (ab < 1e-5f || cb < 1e-5f) return null
            val dot = (a.x-b.x)*(c.x-b.x) + (a.y-b.y)*(c.y-b.y) + (a.z-b.z)*(c.z-b.z)
            return acos((dot / (ab * cb)).coerceIn(-1f, 1f)) / PI.toFloat()
        }

        fun base(frame: PoseFrame): FloatArray? {
            val w = frame.worldLandmarks; val n = frame.landmarks
            if (requiredJoints.any { min(w[it].visibility, n[it].visibility) < 0.5f }) return null
            val hip = midpoint(w[23], w[24]); val shoulder = midpoint(w[11], w[12])
            val torso = distance(hip, shoulder)
            if (torso < 1e-4f) return null
            val nh = midpoint(n[23], n[24]); val ns = midpoint(n[11], n[12])
            val aspect = frame.imageWidth.toFloat() / frame.imageHeight
            val scale = sqrt(((ns.x-nh.x)*aspect).pow(2) + (ns.y-nh.y).pow(2))
            if (scale < 1e-4f) return null
            val angles = listOf(Triple(23,25,27), Triple(24,26,28), Triple(11,23,25), Triple(12,24,26), Triple(11,13,15), Triple(12,14,16))
                .map { (a,b,c) -> angle(w[a],w[b],w[c]) ?: return null }
            val result = FloatArray(BASE_COUNT)
            angles.forEachIndexed { i, value -> result[i] = value }
            result[6] = acos(((hip.y - shoulder.y) / torso).coerceIn(-1f, 1f)) / PI.toFloat()
            result[7] = distance(w[15], hip) / torso
            result[8] = distance(w[16], hip) / torso
            result[9] = distance(w[27], w[28]) / torso
            result[10] = nh.y
            result[11] = (n[27].y + n[28].y) / 2
            result[12] = scale
            result[13] = ((w[15].y + w[16].y) / 2 - hip.y) / torso
            result[14] = ((w[27].y + w[28].y) / 2 - hip.y) / torso
            result[15] = distance(hip, midpoint(w[27], w[28])) / torso
            return result.takeIf { it.all(Float::isFinite) }
        }
    }
}

/** Causal sample-and-hold: never interpolates from a future frame. Holds at most 100 ms. */
class TemporalWindow(private val manifest: ModelManifest) {
    private val rows = ArrayDeque<FloatArray>()
    private var origin: Long? = null
    private var tick = 0L
    private var previousTime = 0L
    private var previous: PoseFeatures? = null

    fun reset() { rows.clear(); origin = null; tick = 0; previous = null }

    fun add(timestampMs: Long, features: PoseFeatures?) {
        if (origin == null) origin = timestampMs
        val start = origin!!
        // A long gap must not allocate/process an unbounded number of empty samples.
        val latestTick = (timestampMs - start) * manifest.sampleRateHz / 1000
        if (latestTick - tick > manifest.windowSize) {
            rows.clear(); tick = latestTick - manifest.windowSize + 1
        }
        while (start + tick * 1000 / manifest.sampleRateHz <= timestampMs) {
            val at = start + tick * 1000 / manifest.sampleRateHz
            val source = if (at == timestampMs) features else previous?.takeIf { at - previousTime <= 100 }
            val row = FloatArray(manifest.featureCount + 1)
            if (source != null) {
                for (i in 0 until manifest.featureCount) row[i] = (source.values[i] - manifest.mean[i]) / manifest.std[i]
                row[manifest.featureCount] = 1f
            }
            rows.addLast(row)
            if (rows.size > manifest.windowSize) rows.removeFirst()
            tick++
        }
        previous = features
        previousTime = timestampMs
    }

    fun snapshot(): FeatureWindow {
        val channels = manifest.featureCount + 1
        val output = FloatArray(manifest.windowSize * channels)
        var offset = (manifest.windowSize - rows.size) * channels
        var count = 0
        rows.forEach { row ->
            row.copyInto(output, offset); offset += channels
            if (row.last() == 1f) count++
        }
        return FeatureWindow(output, manifest.windowSize, channels, count)
    }
}

File: training/motion_training/features.py
"""Keep arithmetic, channel order and causal resampling in sync with core/Features.kt."""
import math
from collections import deque

CLASSES = ["squat", "burpee", "jump_squat", "push_up", "other"]
FEATURE_COUNT = 32
WINDOW_SIZE = 180
SAMPLE_RATE = 30
REQUIRED = (11, 12, 13, 14, 15, 16, 23, 24, 25, 26, 27, 28)


def validate_pose(frame):
    if not isinstance(frame["timestampMs"], int) or frame["timestampMs"] < 0:
        raise ValueError("Invalid timestamp")
    if frame["imageWidth"] <= 0 or frame["imageHeight"] <= 0:
        raise ValueError("Invalid image dimensions")
    for name in ("landmarks", "worldLandmarks"):
        if len(frame[name]) != 33:
            raise ValueError("Expected 33 landmarks")
        for p in frame[name]:
            if not all(math.isfinite(p[k]) for k in ("x", "y", "z")) or not 0 <= p.get("visibility", 1) <= 1:
                raise ValueError("Invalid coordinate/visibility")


def midpoint(a, b):
    return {k: (a[k] + b[k]) / 2 for k in ("x", "y", "z")}


def distance(a, b):
    return math.sqrt(sum((a[k] - b[k]) ** 2 for k in ("x", "y", "z")))


def angle(a, b, c):
    ab, cb = distance(a, b), distance(c, b)
    if min(ab, cb) < 1e-5:
        return None
    dot = sum((a[k] - b[k]) * (c[k] - b[k]) for k in ("x", "y", "z"))
    return math.acos(max(-1, min(1, dot / (ab * cb)))) / math.pi


def base_features(frame):
    validate_pose(frame)
    w, n = frame["worldLandmarks"], frame["landmarks"]
    if any(min(w[i].get("visibility", 1), n[i].get("visibility", 1)) < .5 for i in REQUIRED):
        return None
    hip, shoulder = midpoint(w[23], w[24]), midpoint(w[11], w[12])
    torso = distance(hip, shoulder)
    nh, ns = midpoint(n[23], n[24]), midpoint(n[11], n[12])
    aspect = frame["imageWidth"] / frame["imageHeight"]
    scale = math.hypot((ns["x"] - nh["x"]) * aspect, ns["y"] - nh["y"])
    if torso < 1e-4 or scale < 1e-4:
        return None
    result = [angle(w[a], w[b], w[c]) for a, b, c in
              ((23, 25, 27), (24, 26, 28), (11, 23, 25), (12, 24, 26), (11, 13, 15), (12, 14, 16))]
    if any(v is None for v in result):
        return None
    result.extend([
        math.acos(max(-1, min(1, (hip["y"] - shoulder["y"]) / torso))) / math.pi,
        distance(w[15], hip) / torso, distance(w[16], hip) / torso,
        distance(w[27], w[28]) / torso, nh["y"], (n[27]["y"] + n[28]["y"]) / 2, scale,
        ((w[15]["y"] + w[16]["y"]) / 2 - hip["y"]) / torso,
        ((w[27]["y"] + w[28]["y"]) / 2 - hip["y"]) / torso,
        distance(hip, midpoint(w[27], w[28])) / torso,
    ])
    return result if all(math.isfinite(v) for v in result) else None


class FeatureExtractor:
    def __init__(self):
        self.previous = None

    def reset(self):
        self.previous = None

    def extract(self, frame):
        base = base_features(frame)
        if base is None:
            self.reset()
            return None
        derivatives = [0.] * 16
        if self.previous:
            last_time, last_values = self.previous
            dt = (frame["timestampMs"] - last_time) / 1000
            if 0 < dt <= .5:
                derivatives = [max(-20, min(20, (base[i] - last_values[i]) / dt)) for i in range(16)]
        self.previous = (frame["timestampMs"], base)
        return base + derivatives


class TemporalWindow:
    def __init__(self, mean=None, std=None):
        self.mean = mean if mean is not None else [0.] * 32
        self.std = std if std is not None else [1.] * 32
        self.reset()

    def reset(self):
        self.rows = deque(maxlen=WINDOW_SIZE)
        self.origin = None
        self.tick = 0
        self.previous = None
        self.previous_time = 0

    def add(self, timestamp, features):
        if self.origin is None:
            self.origin = timestamp
        latest_tick = (timestamp - self.origin) * SAMPLE_RATE // 1000
        if latest_tick - self.tick > WINDOW_SIZE:
            self.rows.clear()
            self.tick = latest_tick - WINDOW_SIZE + 1
        while self.origin + self.tick * 1000 // SAMPLE_RATE <= timestamp:
            at = self.origin + self.tick * 1000 // SAMPLE_RATE
            source = features if at == timestamp else self.previous if at - self.previous_time <= 100 else None
            row = [(source[i] - self.mean[i]) / self.std[i] for i in range(32)] + [1.] if source is not None else [0.] * 33
            self.rows.append(row)
            self.tick += 1
        self.previous, self.previous_time = features, timestamp

    def snapshot(self):
        return [[0.] * 33 for _ in range(WINDOW_SIZE - len(self.rows))] + list(self.rows)


def feature_stream(records):
    """Same gap reset and invalid-pose derivative handling as MotionEngine."""
    extractor = FeatureExtractor()
    window = TemporalWindow()
    last_good = None
    for record in records:
        timestamp = record["timestampMs"]
        if last_good is not None and timestamp - last_good > 500:
            extractor.reset()
            window.reset()
            last_good = None
        pose = record.get("pose")
        values = extractor.extract(pose) if pose is not None else None
        if pose is None:
            extractor.reset()
        if values is not None:
            last_good = timestamp
        window.add(timestamp, values)
        yield timestamp, values, window.snapshot()
